File: apps/facilitator/packages/python/x402/src/fourmica_x402/facilitator.py
"""4mica facilitator client wrappers for the x402 Python SDK."""

from __future__ import annotations

from typing import Any, Dict

from x402.http import (
    FacilitatorConfig,
    HTTPFacilitatorClient,
    HTTPFacilitatorClientSync,
)
from x402.schemas import PaymentRequirements, SettleResponse
from x402.schemas.v1 import PaymentRequirementsV1

DEFAULT_FOURMICA_FACILITATOR_URL = "https://x402.4mica.xyz"


def _model_to_payload(model: Any) -> Dict[str, Any]:
    if hasattr(model, "model_dump"):
        return model.model_dump(by_alias=True, exclude_none=True)
    if hasattr(model, "dict"):
        return model.dict(by_alias=True, exclude_none=True)
    if isinstance(model, dict):
        return model
    raise TypeError("payload must be a dict or pydantic model")


class FourMicaFacilitatorClient(HTTPFacilitatorClient):
    """Async 4mica facilitator client."""

    def __init__(self, config: FacilitatorConfig | dict[str, Any] | None = None) -> None:
        if isinstance(config, dict):
            config = {"url": config.get("url", DEFAULT_FOURMICA_FACILITATOR_URL), **config}
        else:
            config = config or FacilitatorConfig(url=DEFAULT_FOURMICA_FACILITATOR_URL)
            if config.url is None:
                config.url = DEFAULT_FOURMICA_FACILITATOR_URL
        super().__init__(config)

    async def settle(
        self,
        payload,
        requirements,
    ) -> SettleResponse:
        request_body = self._build_request_body(
            payload.x402_version,
            _model_to_payload(payload),
            _model_to_payload(requirements),
        )
        client = self._get_async_client()
        response = await client.post(
            f"{self._url}/settle",
            headers=self._get_settle_headers(),
            json=request_body,
        )
        if response.status_code != 200:
            raise ValueError(f"Facilitator settle failed ({response.status_code}): {response.text}")
        return _normalize_settle_response(response.json(), requirements)


class FourMicaFacilitatorClientSync(HTTPFacilitatorClientSync):
    """Sync 4mica facilitator client."""

    def __init__(self, config: FacilitatorConfig | dict[str, Any] | None = None) -> None:
        if isinstance(config, dict):
            config = {"url": config.get("url", DEFAULT_FOURMICA_FACILITATOR_URL), **config}
        else:
            config = config or FacilitatorConfig(url=DEFAULT_FOURMICA_FACILITATOR_URL)
            if config.url is None:
                config.url = DEFAULT_FOURMICA_FACILITATOR_URL
        super().__init__(config)

    def settle(
        self,
        payload,
        requirements,
    ) -> SettleResponse:
        request_body = self._build_request_body(
            payload.x402_version,
            _model_to_payload(payload),
            _model_to_payload(requirements),
        )
        client = self._get_client()
        response = client.post(
            f"{self._url}/settle",
            headers=self._get_settle_headers(),
            json=request_body,
        )
        if response.status_code != 200:
            raise ValueError(f"Facilitator settle failed ({response.status_code}): {response.text}")
        return _normalize_settle_response(response.json(), requirements)


def _attach_optional_fields(
    response: SettleResponse,
    payload: Dict[str, Any],
) -> SettleResponse:
    certificate = payload.get("certificate")
    transaction = (
        payload.get("transaction")
        or payload.get("transactionHash")
        or payload.get("txHash")
        or payload.get("tx_hash")
        or payload.get("hash")
        or payload.get("requestId")
        or payload.get("request_id")
    )
    tx_hash = payload.get("txHash") or payload.get("tx_hash")
    network = (
        payload.get("network")
        or payload.get("networkId")
        or payload.get("network_id")
        or payload.get("chainId")
        or payload.get("chain_id")
    )
    network_id = payload.get("networkId") or payload.get("network_id")
    payer = payload.get("payer") or payload.get("userAddress") or payload.get("user_address")

    for name, value in (
        ("certificate", certificate),
        ("transaction", transaction),
        ("tx_hash", tx_hash),
        ("txHash", tx_hash),
        ("network", network),
        ("network_id", network_id),
        ("networkId", network_id),
        ("payer", payer),
    ):
        if value is None:
            continue
        try:
            setattr(response, name, value)
        except Exception:
            continue
    return response


def _normalize_settle_response(
    payload: Dict[str, Any],
    requirements: PaymentRequirements | PaymentRequirementsV1,
) -> SettleResponse:
    try:
        response = SettleResponse.model_validate(payload)
    except Exception:
        if not isinstance(payload, dict):
            raise
    else:
        return _attach_optional_fields(response, payload)

    tx = (
        payload.get("transaction")
        or payload.get("transactionHash")
        or payload.get("txHash")
        or payload.get("tx_hash")
        or payload.get("hash")
        or payload.get("requestId")
        or payload.get("request_id")
    )
    network = (
        payload.get("network")
        or payload.get("networkId")
        or payload.get("network_id")
        or payload.get("chainId")
        or payload.get("chain_id")
        or str(requirements.network)
    )
    error_reason = (
        payload.get("error_reason")
        or payload.get("errorReason")
        or payload.get("error")
        or payload.get("message")
    )
    error_message = payload.get("error_message") or payload.get("errorMessage")
    payer = payload.get("payer") or payload.get("userAddress") or payload.get("user_address")
    success = bool(payload.get("success", error_reason is None))

    response = SettleResponse(
        success=success,
        error_reason=error_reason,
        error_message=error_message,
        payer=payer,
        transaction=str(tx or ""),
        network=str(network),
    )
    return _attach_optional_fields(response, payload)

File: apps/facilitator/packages/typescript/x402/src/server/facilitator.ts
import { type FacilitatorConfig, HTTPFacilitatorClient } from '@x402/core/server'
import type { Network, PaymentPayload, PaymentRequirements, SettleResponse } from '@x402/core/types'

const DEFAULT_FACILITATOR_URL = 'https://x402.4mica.xyz'

export interface CertificateResponse {
  claims: string
  signature: string
}

export type FourMicaSettleResponse = SettleResponse & {
  certificate?: CertificateResponse
  txHash?: string
  networkId?: string
  error?: string
}

export class FourMicaFacilitatorClient extends HTTPFacilitatorClient {
  constructor(config?: FacilitatorConfig) {
    super({ ...config, url: config?.url ?? DEFAULT_FACILITATOR_URL })
  }

  async settle(
    paymentPayload: PaymentPayload,
    paymentRequirements: PaymentRequirements
  ): Promise<FourMicaSettleResponse> {
    let headers: Record<string, string> = {
      'Content-Type': 'application/json',
    }

    const authHeaders = await this.createAuthHeaders('settle')
    headers = { ...headers, ...authHeaders.headers }

    const response = await fetch(`${this.url}/settle`, {
      method: 'POST',
      headers,
      body: JSON.stringify(
        this.safeJson({
          x402Version: paymentPayload.x402Version,
          paymentPayload,
          paymentRequirements,
        })
      ),
    })

    const data = (await response.json()) as Record<string, unknown>
    const normalized = normalizeSettleResponse(data, paymentRequirements)

    if (!response.ok || !normalized.success) {
      throw new Error(
        `Facilitator settle failed (${response.status}): ${normalized.errorReason ?? normalized.error ?? 'unknown error'}`
      )
    }

    return normalized
  }

  /**
   * Helper to convert objects to JSON-safe format.
   * Handles BigInt and other non-JSON types.
   *
   * @param obj - The object to convert
   * @returns The JSON-safe representation of the object
   */
  private safeJson<T>(obj: T): T {
    return JSON.parse(
      JSON.stringify(obj, (_, value) => (typeof value === 'bigint' ? value.toString() : value))
    )
  }
}

function normalizeSettleResponse(
  payload: Record<string, unknown>,
  requirements: PaymentRequirements
): FourMicaSettleResponse {
  const transaction = String(
    payload.transaction ??
      payload.transactionHash ??
      payload.txHash ??
      payload.tx_hash ??
      payload.hash ??
      ''
  )
  const network = String(
    payload.network ?? payload.networkId ?? payload.network_id ?? requirements.network
  ) as Network
  const errorReason =
    typeof payload.errorReason === 'string'
      ? payload.errorReason
      : typeof payload.error_reason === 'string'
        ? payload.error_reason
        : typeof payload.error === 'string'
          ? payload.error
          : typeof payload.message === 'string'
            ? payload.message
            : undefined

  const certificate =
    payload.certificate &&
    typeof payload.certificate === 'object' &&
    typeof (payload.certificate as Record<string, unknown>).claims === 'string' &&
    typeof (payload.certificate as Record<string, unknown>).signature === 'string'
      ? {
          claims: (payload.certificate as Record<string, string>).claims,
          signature: (payload.certificate as Record<string, string>).signature,
        }
      : undefined

  return {
    success: Boolean(payload.success ?? errorReason === undefined),
    errorReason,
    payer:
      typeof payload.payer === 'string'
        ? payload.payer
        : typeof payload.userAddress === 'string'
          ? payload.userAddress
          : typeof payload.user_address === 'string'
            ? payload.user_address
            : undefined,
    transaction,
    network,
    certificate,
    txHash:
      typeof payload.txHash === 'string'
        ? payload.txHash
        : typeof payload.tx_hash === 'string'
          ? payload.tx_hash
          : transaction || undefined,
    networkId:
      typeof payload.networkId === 'string'
        ? payload.networkId
        : typeof payload.network_id === 'string'
          ? payload.network_id
          : network,
    error: typeof payload.error === 'string' ? payload.error : undefined,
  }
}
